Client/TabManagement/TabManagerImplementation.swift

```diff
@@ -467,3 +475,76 @@ class TabManagerImplementation: LegacyTabManager, Notifiable {
         }
     }
 }
+
+extension TabManagerImplementation {
+    // Ecosia: Get Sites from places
+    private func fetchDBSites() async -> [Site] {
+        do {
+            let sites = try await withCheckedThrowingContinuation { continuation in
+                profile.places.getSitesWithBound(
+                    limit: 100,
+                    offset: 0,
+                    excludedTypes: VisitTransitionSet(0)
+                ).upon { result in
+                    if let successValue = result.successValue {
+                        continuation.resume(returning: successValue.asArray())
+                    } else {
+                        continuation.resume(returning: [])
+                    }
+                }
+            }
+            return sites
+        } catch {
+            return []
+        }
+    }
+    
+    // Ecosia: Fix tabs that where migrated from v9.x to v10.0.0 by fetching from places DB instead.
+    // This can be removed after there are no significant number of users on v10.0.0.
+    private static let restoreMigratedv10TabsMissingUrlKey = "restoreMigratedv10TabsMissingUrl"
+    private var shouldRestoreMigratedv10TabsMissingUrl: Bool {
+        !UserDefaults.standard.bool(forKey: Self.restoreMigratedv10TabsMissingUrlKey)
+    }
+    private func restoreMigratedv10TabsMissingUrlIfNeeded(window: WindowData?) async -> WindowData? {
+        defer {
+            UserDefaults.standard.setValue(true, forKey: Self.restoreMigratedv10TabsMissingUrlKey)
+        }
+        // We only want to run it if the user just upgraded to the version containing this restoration code.
+        // Otherwise it means the user is a fresh install and we just mark it as restorarion done.
+        // This also covers the case where the user migrated directly to >= v10.0.3 since in that case the first run of the app will `TabMigrationUtility.runMigration` and therefore not reach here.
```

Comment:
@d4r1091 I guess this is a plus, right? Just realised it and also added to the comment.

Reply:
indeed some graned logic to minimize the impact on other scenarios